osbuild/__init__.py

```diff
@@ -241,40 +385,58 @@ def add_stage(self, name, options=None):
     def set_assembler(self, name, options=None):
         self.assembler = Assembler(name, options or {})
 
-    def run(self, output_dir, objects=None, interactive=False, check=True, libdir=None):
+    def run(self, output_dir, store, interactive=False, check=True, libdir=None):
         os.makedirs("/run/osbuild", exist_ok=True)
-        if objects:
-            os.makedirs(objects, exist_ok=True)
-        elif self.base:
-            raise ValueError("'objects' argument must be given when pipeline has a 'base'")
-
+        if self.base and not store:
+            raise ValueError("'store' argument must be given when pipeline has a 'base'")
+        object_store = ObjectStore(store)
         results = {
             "stages": []
         }
-        with TmpFs() as tree:
-            if self.base:
-                subprocess.run(["cp", "-a", f"{objects}/{self.base}/.", tree], check=True)
-
-            for stage in self.stages:
-                r = stage.run(tree, interactive, check, libdir=libdir)
-                results["stages"].append(r)
-                if r["returncode"] != 0:
-                    results["returncode"] = r["returncode"]
-                    return results
-
-            if self.assembler:
-                r = self.assembler.run(tree, output_dir, interactive, check, libdir=libdir)
-                results["assembler"] = r
-                if r["returncode"] != 0:
-                    results["returncode"] = r["returncode"]
-                    return results
-
-            last = self.stages[-1].id if self.stages else self.base
-            if objects and last:
-                output_tree = f"{objects}/{last}"
-                shutil.rmtree(output_tree, ignore_errors=True)
-                os.makedirs(output_tree, mode=0o755)
-                subprocess.run(["cp", "-a", f"{tree}/.", output_tree], check=True)
+        tree_id = self.stages[-1].id if self.stages else None
+        with object_store.get_tree(tree_id) as existing_tree:
+            if existing_tree:
```

Comment:
This control flow is hard to grasp and duplicates the assembler run. May I suggest something like this:
```python
if not store.has_tree(tree_id):
    with store.new_tree(tree_id, base_id):
        for stage in stages:
            # run stages to generate `tree_id`

if self.assembler:
    with store.get_tree(tree_id):
        # run assembler
```
This would read easier if we could skip a `with` block in certain situations, but that's not possible (without hacks).

Hope I'm not missing something obvious here…

Reply:
You may. And I will.

The reason for the supreme awkwardness, was that i did not want to check for existence before calling `get_tree()`. But that's out of habit, it does not make sense here.

We anage the store, and we make sure that all objects as well as refs remain valid "forever". The only theoretical issue is that with the old logic we would not have a race condition against someone deleting a ref (as long as they don't delete the object). But there is no API for deleting refs, so protecting against that makes little sense when we don't protect against objects being deleted.

Reply:
Makes sense. Thanks!